File: frontend/src/api/ifrsReportHistory.ts
/**
 * IFRS report history API — даты публикации МСФО-отчётности по компаниям (с 2022).
 */
import { api } from "./client";

export interface IfrsHistoryRow {
  company_id: string;
  year: number;
  published_on: string | null;   // 'YYYY-MM-DD'
  updated_by_name: string | null;
  updated_at: string | null;
}

export interface IfrsHistoryLastChange {
  by_name: string | null;
  at: string | null;
}

export interface IfrsHistoryResponse {
  rows: IfrsHistoryRow[];
  last_change: IfrsHistoryLastChange;
}

export const ifrsReportHistoryApi = {
  async list(): Promise<IfrsHistoryResponse> {
    const r = await api.get<IfrsHistoryResponse>("/ifrs-report-history");
    return r.data;
  },
  async upsert(companyId: string, year: number, publishedOn: string | null): Promise<IfrsHistoryRow> {
    const r = await api.put<IfrsHistoryRow>(`/ifrs-report-history/${companyId}/${year}`, {
      published_on: publishedOn,
    });
    return r.data;
  },
};

File: backend/app/schemas/ifrs_report_history.py
"""IFRS report history schemas — даты публикации МСФО-отчётности."""
from __future__ import annotations

from datetime import date, datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class IfrsHistoryRow(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    company_id: UUID
    year: int
    published_on: Optional[date] = None
    updated_by_name: Optional[str] = None
    updated_at: Optional[datetime] = None


class IfrsHistoryLastChange(BaseModel):
    by_name: Optional[str] = None
    at: Optional[datetime] = None


class IfrsHistoryResponse(BaseModel):
    rows: list[IfrsHistoryRow]
    last_change: IfrsHistoryLastChange


class IfrsHistoryUpsert(BaseModel):
    published_on: Optional[date] = None   # null = очистить дату
